pkg/agent/multicast/mcast_controller.go

```diff
@@ -0,0 +1,362 @@
+
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	// Periodically query Multicast Groups on OVS.
+	go wait.NonSlidingUntil(func() {
+		if err := c.igmpSnooper.queryIGMP(net.IPv4zero, queryVersion); err != nil {
+			klog.ErrorS(err, "Failed to send IGMP query")
+		}
+	}, queryInterval, stopCh)
+
+	// Periodically check the group member status, and remove the groups in which no members exist
+	go wait.NonSlidingUntil(c.clearStaleGroups, queryInterval, stopCh)
+
+	go c.eventHandler(stopCh)
+
+	for i := 0; i < int(workerCount); i++ {
+		// Process multicast Group membership report or leave messages.
+		c.worker()
+	}
+}
+
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// getGroupMemberStatusByGroup returns the GroupMemberStatus according to the given group.
+func (c *Controller) getGroupMemberStatusByGroup(group net.IP) *GroupMemberStatus {
+	status, ok, _ := c.groupCache.GetByKey(group.String())
+	if ok {
+		return status.(*GroupMemberStatus)
+	}
+	return nil
+}
+
+// getGroupMemberStatusesByPod returns all GroupMemberStatus that the given podInterface is included in its localMembers.
+func (c *Controller) getGroupMemberStatusesByPod(podInterface string) []*GroupMemberStatus {
+	groupMembers := make([]*GroupMemberStatus, 0)
+	statuses, _ := c.groupCache.ByIndex(podInterfaceIndex, podInterface)
+	for _, s := range statuses {
+		groupMembers = append(groupMembers, s.(*GroupMemberStatus))
+	}
+	return groupMembers
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	// We expect string (multicast group) to come off the workqueue.
+	if key, ok := obj.(string); !ok {
+		// As the item in the workqueue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncGroup(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.Errorf("Error syncing multicast group %s, requeuing. Error: %v", key, err)
+	}
+	return true
+
+}
+
+func (c *Controller) syncGroup(groupKey string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing Egress for %s. (%v)", groupKey, time.Since(startTime))
+	}()
+	obj, exists, err := c.groupCache.GetByKey(groupKey)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get GroupMemberStatus", "groupKey", groupKey)
+		return err
+	}
+	if !exists {
+		klog.InfoS("multicast groupKey not found in the cache", "groupKey", groupKey)
+		return nil
+	}
+	status := obj.(*GroupMemberStatus)
+	if c.groupHasInstalled(groupKey) {
+		if c.groupIsStale(status) {
```

Comment:
I wonder what if `clearStaleGroups` dicovers a stale group, could the group be removed before `checkLastMember` is executed once? What's the condition that a group can be removed? based on probe time or the number of members?